pkg/util/crypt/crypt_dev.go

```diff
@@ -0,0 +1,306 @@
+
+	// Open a new Temp file to stash the loop contents
+	loopF, err := ioutil.TempFile("", "loop-")
+	if err != nil {
+		sylog.Debugf("Error creating temporary crypt file")
+		return "", "", err
+	}
+
+	copyDeviceContents("/dev/"+loopdev, loopF.Name(), size)
+
+	return loopF.Name(), nextCrypt, err
+}
+
+// copyDeviceContents copies the contents of source to destination.
+// source and dest can either be a file or a block device
+func copyDeviceContents(source string, dest string, size int64) error {
+
+	squashFd, err := syscall.Open(source, syscall.O_RDWR, 0777)
+	if err != nil {
+		sylog.Debugf("Unable to open the file %s", source)
+		return err
+	}
+	defer syscall.Close(squashFd)
+
+	cryptFd, err := syscall.Open(dest, syscall.O_RDWR, 0777)
+
+	if err != nil {
+		sylog.Debugf("Unable to open the file: %s", dest)
+		return err
+	}
+
+	defer syscall.Close(cryptFd)
+
+	var writtenSoFar int64
+
+	buffer := make([]byte, 1024)
+	for writtenSoFar < size {
+		numRead, err := syscall.Read(squashFd, buffer)
+		if err != nil {
+			sylog.Debugf("Unable to read the the file %s", source)
+			return err
+		}
+		numWritten, err := syscall.Write(cryptFd, buffer)
+		if err != nil {
+			sylog.Debugf("Unable to write to destination %s", dest)
+			return err
+		}
+
+		if numRead != numWritten {
+			sylog.Debugf("numRead != numWritten")
+		}
+		writtenSoFar += int64(numWritten)
+	}
+
+	return nil
+}
+
+func getRandomString(n int) string {
+	b := make([]byte, n)
+	if _, err := rand.Read(b); err != nil {
+		sylog.Debugf("Something went wrong while generating random string")
+		return ""
+	}
+	return fmt.Sprintf("%x", b)
+}
+
+func getNextAvailableCryptDevice() string {
+	return getRandomString(15)
+}
+
+// GetCryptDevice returns the next available device in /dev/mapper for encryption/decryption
+func (crypt *Device) GetCryptDevice(loopDev string) (string, error) {
+
+	fmt.Print("Enter the password to decrypt the File System: ")
+	password, err := terminal.ReadPassword(int(syscall.Stdin))
+	if err != nil {
+		sylog.Fatalf("Error parsing input: %s", err)
+	}
+	fmt.Println()
+
+	fd, err := lock.Exclusive("/dev/mapper")
+	if err != nil {
+		sylog.Debugf("Unable to acquire lock on /dev/mapper while decrypting")
+		return "", err
+	}
+	defer lock.Release(fd)
+
+	maxRetries := 3 // Arbitrary number of retries.
+
+retry:
+	numRetries := 0
+	nextCrypt := getNextAvailableCryptDevice()
+	if nextCrypt == "" {
+		return "", errors.New("Crypt Device not available")
+	}
+
+	cmd := exec.Command("/sbin/cryptsetup", "luksOpen", loopDev, nextCrypt)
+	cmd.Dir = "/dev"
+	cmd.SysProcAttr = &syscall.SysProcAttr{}
+	cmd.SysProcAttr.Credential = &syscall.Credential{Uid: 0, Gid: 0}
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, string(password))
+	}()
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		if strings.Contains(string(out), "No key available") {
```

Comment:
Just a question, is cryptsetup return different exit code which could help to determine the error ? Relying on error output could have side effects between cryptsetup version, `LANG` environment variable can alter error output and return error message in another language, ok it's not actually the case because this function is only called by RPC server which run with only one environment variable, but if this function is called directly by build command or something else, that's something to consider.

Reply:
cryptsetup error codes are not as descriptive as the output. That's why I was checking for the output to know what exactly the error is. Here are cryptsetup error codes.

0 - success, otherwise fail
1 - wrong parameters
2 - no permission
3 - out of memory
4 - wrong device specified
5 - device already exists or device is busy

If the password is wrong or for any other wrong parameter, the error code would be 1. For the latter case where the device exists, the code would be 5. I can replace the second string compare by checking the error code for 5 but error code 1 is not very descriptive.